src/aframe_components/custom-components.ts

```diff
@@ -20,24 +25,100 @@ interface State {
 export interface Guidance extends Component {
   state: State;
   pointPositions: Vector3[];
+  kdtree: KDNode | null;
+}
+
+export interface KDNode {
+  position: Vector3;
+  left: KDNode | null;
+  right: KDNode | null;
+}
+
+export function build3DTree(positions: Vector3[], depth: number){
+  const n = positions.length;
+  const mid = Math.floor(n / 2);
+  if (n <= 0){
+    return null;
+  }
+  const axis = depth % 3;
+  positions.sort((a, b) => a[AXIS_DICT[axis]] - b[AXIS_DICT[axis]]);
+  return {
+    position: positions[mid],
+    left: build3DTree(positions.slice(0, mid), depth + 1),
```

Comment:
Instead of slicing the array at every step, you could pass a start and end index into `build3DTree` and operate on only those.  This way you're not allocating two arrays each time you recurse